File: Leetcode Challenge/February/cpp/Contiguous Array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int,int> mp; 
        
        
        for(int i = 0; i < nums.size(); i++) {
            if(nums[i] == 0) {
                nums[i] = -1;
            }
        }
        
        int sum=0,res=0;
        for(int i=0;i<nums.size();i++){
            sum+=nums[i];
            
            if(sum == 0){
                res = max(res, i+1);
            }
            
            if(mp.find(sum)!=mp.end()){ //sum is in map
                res=max(res,i-mp[sum]);
            }
            else{
                mp[sum]=i;
            }
        }
        return res;
    }
};

File: Leetcode Challenge/February/java/Contiguous Array.java
class Solution {
    public int findMaxLength(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        for(int i = 0; i < nums.length; i++){
            if(nums[i] == 0) nums[i] = -1;
        }
        int sum = 0; 
        int max = 0; 
        Map<Integer, Integer> map = new HashMap<>();
        
        for(int i = 0; i < nums.length; i++){
            sum += nums[i]; 
            if(sum == 0) {
                max = Math.max(max, i+1);
            }
            if(map.containsKey(sum)){ 
                int last = map.get(sum); 
                max = Math.max(max, i - last); 
            }
            else{ 
                map.put(sum, i);
            }
        }
        return max;
    }
}
